### PR TITLE
ConcordClient: Self-contained API

### PR DESCRIPTION
**Pull request by robem**
This PR breaks the dependency between request service and concord client pool.
In order to keep the separation between our public client interface (bft client, concord client, clientservice) clean, users should not need to depend on implementation details such as the client pool or the thin replica client.

Please take a look at the individual commit messages for more details.

### CODE REVIEW

File: bftengine/include/bftengine/SimpleClient.hpp [hunk 1]
```diff
@@ -70,7 +66,7 @@ struct ClientReply {
   OperationResult opResult = SUCCESS;
   std::string cid;
   std::string span_context;
-  RequestCallBack cb = {};
+  concord::client::concordclient::SendCallback cb = {};
```
[the remark this answers is not in the document]
Comment (robem) [reply bftengine/include/bftengine/SimpleClient.hpp]: I doesn't. Please check the last commit.

File: client/concordclient/include/client/concordclient/concord_client.hpp [hunk 2]
```diff
@@ -148,18 +145,8 @@ class ConcordClient {
...
-  const ConcordClientConfig& config_;
-  std::shared_ptr<concordMetrics::Aggregator> metrics_;
-
-  // TODO: Allow multiple subscriptions
-  std::atomic_bool active_subscription_{false};
-
-  std::shared_ptr<BasicUpdateQueue> trc_queue_;
-  std::unique_ptr<::client::thin_replica_client::ThinReplicaClient> trc_;
-  std::unique_ptr<concord::concord_client_pool::ConcordClientPool> client_pool_;
+  class Impl;
```
[the remark this answers is not in the document]
Comment (robem) [reply client/concordclient/include/client/concordclient/concord_client.hpp:140]: This class is defined inside the ConcordClient class - meaning it is `ConcordClient::Impl`. The class itself is supposed to only be used inside the ConcordClient class due to it includes the implementation details.
Do you have any suggestions? `ConcordClient::ConcordClientImpl`?

File: bftengine/include/bftengine/SimpleClient.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuliasherman) [reply bftengine/include/bftengine/SimpleClient.hpp]: Yes, I saw it later, thanks!

File: client/concordclient/include/client/concordclient/concord_client.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuliasherman) [reply client/concordclient/include/client/concordclient/concord_client.hpp:140]: Yes, ConcordClientImpl is a good choice.

**Review by yuliasherman — APPROVED** [review]
LGTM, 10x.

File: client/client_pool/include/client/client_pool/concord_client_pool.hpp [hunk 3]
```diff
@@ -124,19 +125,19 @@ class ConcordClientPool {
...
                            std::string span_context = std::string(),
-                           const bftEngine::RequestCallBack& callback = {});
+                           const concord::client::concordclient::SendCallback& callback = {});
 
   // This method is responsible to get write requests with the new client
   // paramters and parse it to the old SimpleClient interface.
   SubmitResult SendRequest(const bft::client::WriteConfig& config,
                            bft::client::Msg&& request,
-                           const bftEngine::RequestCallBack& callback = {});
+                           const concord::client::concordclient::SendCallback& callback = {});
```
Comment (yuliasherman) [inline client/client_pool/include/client/client_pool/concord_client_pool.hpp:134]: Why do we pass a callback in every SendRequest call and not set it once?

File: client/client_pool/src/concord_client_pool.cpp [hunk 4]
```diff
@@ -42,9 +48,9 @@ SubmitResult ConcordClientPool::SendRequest(std::vector<uint8_t> &&request,
                                             uint64_t seq_num,
                                             std::string correlation_id,
                                             std::string span_context,
-                                            const bftEngine::RequestCallBack &callback) {
+                                            const SendCallback &callback) {
   if (callback && timeout_ms.count() == 0) {
-    callback(bftEngine::SendResult{SubmitResult::InvalidArgument});
+    callback(SendResult{SendError::InvalidArgument});
```
Comment (yuliasherman) [inline client/client_pool/src/concord_client_pool.cpp:53]: It's a pity that we are not working in synch. We are changing the approach to the end-to-end error handling: https://confluence.eng.vmware.com/display/BLOC/BFT+client+error+handling+-+one+pager

File: client/client_pool/include/client/client_pool/concord_client_pool.hpp (hunk 3 above)
Comment (robem) [reply client/client_pool/include/client/client_pool/concord_client_pool.hpp:134]: Good point. This PR deals with replacing the callback type. Refactoring the `SendRequest` can be a follow-up.

**Comment by robem** [comment]
The last push was a rebase to incorporate Yulia's latest changes.

**Comment by robem** [comment]
And another rebase. I'm testing the change in closed-source as well which is why it takes a bit longer.

**Comment by robem** [comment]
Closed in favor of #2583

**Pull request closed without merging**